### DISCUSSION TITLE
[Issue with running NeoPixelBitmap.ino on a 8*32 flex display]

### DISCUSSION THREAD

**Discussion by techtoys** [Makuna/NeoPixelBus#738] [Q&A] [answered]
Hi

I am trying to run NeoPixelBitmap.ino sketch on my hardware with configurations below:

NeoPixel is a sheet of 8*32 WS2812B
Host MCU: ESP32S3
IDE: Arduino 2.2.1

Using a very small bitmap of 8*8 pixels to make thing simple. 
The following modifications are required to get the example work:

```
typedef NeoGrbFeature MyPixelColorFeature;
const uint16_t PixelCount = 8*8; // not 8, it is 8*8
NeoPixelBus<MyPixelColorFeature, NeoEsp32BitBangWs2812xMethod> strip(PixelCount, PixelPin);
//...
void LoopAnimUpdate(const AnimationParam& param)
{
    if (param.state == AnimationState_Completed)
    {
        animations.RestartAnimation(param.index);
        //the second argument is not 0. I needed to use  image.Width()*animState to make it work;
        //otherwise, only the first row got refreshed
        image.Blt(strip, image.Width()*animState, 0, animState, image.Width());
        animState = (animState + 1) % image.Height(); // increment and wrap
    }
}

```

Although the bitmap is displayed, it is somewhat distorted. It seems to be an endianness issue. 
Attached please find the original file (sq8x8.png) and the actual result (actual.jpg) as a reference. Please notice that because *.bmp file is not supported by the markdown here, I have converted it to sq8x8.bmp to sq8x8.png.

![sq8x8](https://github.com/Makuna/NeoPixelBus/assets/6328296/f0d8e010-0bba-4c4a-af2f-a53ba7b35b59)
![actual](https://github.com/Makuna/NeoPixelBus/assets/6328296/bd118be5-2022-4ad8-b476-7e478425b99b)

Am I doing it right? How to correct the distortion?

I also wish to be able to control the brightness. How to do it with NepBitmapFile object?

John

**Comment by Makuna** [comment, marked as answer]
(NOT AN ISSUE attribute means it's a discussion, so I moved it over.  If we find a real bug here, we can create an issue to track it with details)  


How did you define `image`?
The actual BMP file is needed to see details of what it is. You can rename the file to .txt and then attach.  Or ZIP it up and attach that.  It looks like the image is rotated 180 degrees which could be from file details.   Or it could be how the physical matrix is mapped out.  What is pixel zero?  What is pixel 7?  What is Pixel 8?  See [Wiki](https://github.com/Makuna/NeoPixelBus/wiki/Matrix-Panels-Support).
You are using BLT method that only does a single row.  see [Wiki](https://github.com/Makuna/NeoPixelBus/wiki/NeoBitmapFile-object-API#void-bltneobuffercontext-destbuffer-uint16_t-indexpixel-int16_t-xsrc-int16_t-ysrc-int16_t-wsrc)
You want to use the one that does 2D Blt, which requires more parameters including the layout MAP for your flexible matrix.  see [Wiki](https://github.com/Makuna/NeoPixelBus/wiki/NeoBitmapFile-object-API#void-bltneobuffercontext-destbuffer-int16_t-xdest-int16_t-ydest-int16_t-xsrc-int16_t-ysrc-int16_t-wsrc-int16_t-hsrc-layoutmapcallback-layoutmap).  
If you want to change the brightness when you copy the pixels, then what you want is to "render" the image not blt it.  This allows for modifications to pixel data.  See [Wiki](https://github.com/Makuna/NeoPixelBus/wiki/NeoBitmapFile-object-API#void-rendershaderneobuffercontextt_color_feature-destbuffer-shader-shader-int16_t-xdest-int16_t-ydest-int16_t-xsrc-int16_t-ysrc-int16_t-wsrc-int16_t-hsrc-layoutmapcallback-layoutmap).  There are some shader examples; even one that does brightness only.

**Reply by techtoys** [reply]
`image ` defined as `NeoBitmapFile<NeoGrbFeature, File> image`.
With NeoTopology declared `NeoTopology<MyPanelLayout> topo(8, 8)` , I can get my 2D matrix board show your `Better.bmp` with function call `image.Blt(strip, 0,0,0,0,8,8, MyLayoutMap)`.

![Better_result](https://github.com/Makuna/NeoPixelBus/assets/6328296/a6e3f8dd-bb76-453c-969f-69ae9c466494)

But I have difficulty with shader. The most relevant example I'm able to locate is 'NeoPixelBufferShader'.
In the example, there is `template<typename T_COLOR_FEATURE> class BrightnessShader : public NeoShaderBase{...}`. 
I am doing the same declaration in my sketch to make a call as  `image.Render<BrightnessShader<NeoGrbFeature>>(strip, shader, 0,0,0,0,8,8, MyLayoutMap)` but the sketch could not compile.

Below is an extract of all essential parts in my sketch:

```
//this part is the same as that in NeoPixelBufferShader.ino
template<typename T_COLOR_FEATURE> class BrightnessShader : public NeoShaderBase
{
public:
  BrightnessShader():
    NeoShaderBase(),
    _brightness(255) // default to full bright
  {}

  // required for a shader object, it will be called for
  // every pixel
  void Apply(uint16_t index, uint8_t* pDest, const uint8_t* pSrc){
...
};

BrightnessShader<NeoGrbFeature> shader;
NeoTopology<MyPanelLayout> topo(8, 8);
NeoPixelBus<NeoGrbFeature, NeoEsp32BitBangWs2812xMethod> strip(8*8, 1);
NeoBitmapFile<NeoGrbFeature, File> image;

uint16_t MyLayoutMap(int16_t x, int16_t y){
  return topo.Map(x, y);
}

setup()
{
  File bitmapFile = SD.open("/Better.bmp");
  image.Begin(bitmapFile);
  shader.setBrightness(64);
  image.Render<BrightnessShader<NeoGrbFeature>>(strip, shader, 0,0,0,0,8,8, MyLayoutMap);
  strip.Show();
}

```

Please forgive as my concept is not clear so there is some mistake.

**Reply by Makuna** [reply]
There are two flavors of shaders, you found the older less supported version.

Look in the NeoPixelDibTest example.  It actually contains an implemented BrightnessShader for RgbColor.

The call would be 
```
image.Render<BrightnessShader>(strip, shader, 0,0,0,0,8,8, MyLayoutMap)
```

**Reply by techtoys** [reply]
It is working. Thank you, Makuna.

**Comment by Makuna** [comment]
Your test image is too symmetrical, here is a modified version of it that will help diagnose issues better.  I included both the PNG and the BMP in the zip file.
[Better.zip](https://github.com/Makuna/NeoPixelBus/files/12795375/Better.zip)

**Comment by techtoys** [comment]
Sorry to have closed it unintentionally. A follow-up question at https://github.com/Makuna/NeoPixelBus/discussions/738#discussioncomment-7189437.
